### ISSUE TITLE
Support conditions

### ISSUE THREAD

**Issue by fabiand** [kubevirt/kubevirt#1454]
/kind enhancement

**What happened**:
`kubectl wait --for condition=…` appeared

**What you expected to happen**:
That we can use kubectl wait with vms (vmi, maybe vm)

**How to reproduce it (as minimally and precisely as possible)**:
`kubectl wait --for condition=ready vmi fedora`

**Anything else we need to know?**:
I think we just need to provide the correct conditions on the VMI object

**Comment by shiywang** [comment]
dibs

**Comment by shiywang** [comment]
@fabiand I have a question about code here https://github.com/kubevirt/kubevirt/blob/master/pkg/virt-controller/watch/vmi.go#L246-L251, why we have to add and remove `condition` on VMI conditionally?

Can we just fully sync up the `condition` between pods and VMIs ?  also the code change should only happen in virt-controller right? no virt-handler change I guess?

**Comment by davidvossel** [comment]
> Can we just fully sync up the condition between pods and VMIs ?

There are Pod conditions that do not directly translate to the VMI.  The "Scheduling" condition happens to mean something for the VMI, which is why it applied here. The "Ready" condition for a Pod doesn't mean anything for a VMI, and would be confusing if we placed it on a VMI because it would indicate that the VMI is actually running and "Ready" when in fact it would just mean that the Pod is up... not necessarily that the qemu process is online within that pod. 

said in another way,  A "Ready" condition for the VMI's pod is not the same thing as a "Ready" condition for the VMI.

> also the code change should only happen in virt-controller right? no virt-handler change I guess?

virt-handler also sets conditions on the VMI.

**Comment by rmohr** [comment]
> said in another way, A "Ready" condition for the VMI's pod is not the same thing as a "Ready" condition for the VMI.

@davidvossel that's correct right now. However, I think that at the end we need to align with the pod ready readiness on the VMI too. If no readiness probe is there, then set the VMI ready after a few seconds, otherwise align with the pod where the readiness probe will be executed. That means that we have to shovel the readiness condition free, so that it can again represent "ready to accept traffic from a service". What do you think?

**Comment by davidvossel** [comment]
> If no readiness probe is there, then set the VMI ready after a few seconds, otherwise align with the pod where the readiness probe will be executed. That means that we have to shovel the readiness condition free, so that it can again represent "ready to accept traffic from a service". What do you think?

@rmohr that's an interesting point about "ready" meaning that the pod (and ultimately the qemu process in our case) is ready to accept traffic.

In KubeVirt, we've been using the readiness checks to mean something different though. A "ready" VMI pod means that the KubeVirt components in that pod are online and ready. This indicates to virt-controller that it is time to schedule the VMI to the node. We are actually depending on this logic to prevent some race conditions from occurring... like ensuring virt-handler will never attempt to start a VMI until we know the VMI's pod environment is ready to accept a VMI spec. 

Ultimately the right thing to do here is to converge the Pod readiness checks to reflect that the VM is ready to accept traffic. As part of that transition, we'll have to make sure we come up with an alternative mechanism to account for the race condition between virt-handler and an uninitialized virt-launcher.

**Comment by fabiand** [comment]
https://github.com/kubernetes/community/blob/master/keps/sig-network/0007-pod-ready%2B%2B.md for some pod readiness context.
